File: frontend/src/api/employers.ts
// ============================================================================
// Organisations + employers API client
// Talks to the Django backend at /learner_api (proxied to :8000 by Vite in dev).
// Backs enrolment."Organisations" and enrolment."Employers" — the two profile
// types created from the Users page's Create menu.
//
// An organisation is a company; an employer is a person at one or more of them.
// The employer form's "Employer Group" picker lists organisations, which is why
// listOrganisations supports search + paging.
// ============================================================================

import type { InvitationOutcome } from '@/pages/users/types';

const ORGS = '/learner_api/organisations';
const EMPLOYERS = '/learner_api/employers';

/** One {day, start, end} row of the organisation form's working-hours repeater. */
export interface WorkingHoursSession {
  day: string;
  start: string;
  end: string;
}

export interface OrganisationRow {
  id: string;
  status: string;
  name: string;
  owner: string;
  category: string;
  /** Shown in the Employer Group picker's own columns. */
  groupType: string;
  parentName: string;
  edrsErnNumber: string;
  apprenticeshipAgreementId: string;
  postCode: string;
  address1: string;
  address2: string;
  cityTown: string;
  county: string;
  country: string;
  workingHours: WorkingHoursSession[];
  contactName: string;
  contactEmail: string;
  contactTelephone: string;
  contactRole: string;
  website: string;
  referenceNumber: string;
  levyPayer: string;
  approxNoOfEmployees: number | null;
  healthAndSafety: string;
  logoUrl: string;
  /** Null when never set — distinct from a deliberate "no". */
  sendHoursVerificationEmails: boolean | null;
}

export interface EmployerRow {
  id: string;
  firstName: string;
  surname: string;
  /** Server-side "first surname", so a list row needs no assembly. */
  name: string;
  gender: string;
  email: string;
  mobile: string;
  postCode: string;
  address1: string;
  address2: string;
  townCity: string;
  county: string;
  country: string;
  /** Organisation ids this person belongs to — the real link. */
  employerGroupIds: string[];
  /** Their names, denormalised server-side so a row renders without a lookup. */
  employerGroupNames: string[];
  /**
   * Present on create only, when the form asked for an invitation. Reports the
   * invitation's fate separately from the employer's creation, which succeeds
   * either way — see login/services.py.
   */
  invitation?: InvitationOutcome;
}

/** Option lists for both forms, served from the constants the API validates against. */
export interface EmployerOptions {
  status: string[];
  groupType: string[];
  levyPayer: string[];
  healthAndSafety: string[];
  owners: string[];
}

export type OrganisationInput = Partial<Omit<OrganisationRow, 'id'>>;
export type EmployerInput = Partial<
  Omit<EmployerRow, 'id' | 'name' | 'employerGroupNames' | 'invitation'>
> & {
  /**
   * Send the "set your password" email on create. Unlike the learner and staff
   * forms this is not stored on the employer record — enrolment."Employers" has
   * no such column, and whether they were invited is answered by
   * login."Invitations". It is an instruction for this request only.
   */
  inviteToPlatform?: boolean;
};

async function request<T>(url: string, init?: RequestInit): Promise<T> {
  let res: Response;
  try {
    res = await fetch(url, {
      // Sends the kbc_session cookie — writes require an authenticated staff
      // session (see login.permissions.staff_only).
      credentials: 'include',
      ...init,
      // Spread last: with `...init` after it, a caller passing any headers at
      // all would silently drop these two, failing the Content-Type parse and
      // the CSRF check.
      headers: {
        'Content-Type': 'application/json',
        'X-Requested-With': 'XMLHttpRequest',
        ...(init?.headers || {}),
      },
    });
  } catch {
    throw new Error('Could not reach the server. Is the backend running on port 8000?');
  }
  const text = await res.text();
  const data = text ? JSON.parse(text) : null;
  if (!res.ok) {
    throw new Error((data && data.error) || `Request failed (${res.status})`);
  }
  return data as T;
}

export interface OrganisationPage {
  count: number;
  page: number;
  pageSize: number;
  results: OrganisationRow[];
}

/**
 * List organisations. `page` drives the Employer Group picker's pager; omitting
 * it returns every row, which is what a plain list screen wants.
 */
export function listOrganisations(opts: { search?: string; status?: string; page?: number } = {}): Promise<OrganisationPage> {
  const qs = new URLSearchParams();
  if (opts.search) qs.set('search', opts.search);
  if (opts.status) qs.set('status', opts.status);
  if (opts.page) qs.set('page', String(opts.page));
  const suffix = qs.toString() ? `?${qs}` : '';
  return request<OrganisationPage>(`${ORGS}/${suffix}`);
}

export function createOrganisation(input: OrganisationInput): Promise<OrganisationRow> {
  return request<OrganisationRow>(`${ORGS}/`, { method: 'POST', body: JSON.stringify(input) });
}

export function updateOrganisation(id: string, patch: OrganisationInput): Promise<OrganisationRow> {
  return request<OrganisationRow>(`${ORGS}/${id}/`, { method: 'PATCH', body: JSON.stringify(patch) });
}

/** List employers, optionally filtered to the people at one organisation. */
export function listEmployers(opts: { search?: string; organisation?: string } = {}): Promise<{ count: number; results: EmployerRow[] }> {
  const qs = new URLSearchParams();
  if (opts.search) qs.set('search', opts.search);
  if (opts.organisation) qs.set('organisation', opts.organisation);
  const suffix = qs.toString() ? `?${qs}` : '';
  return request<{ count: number; results: EmployerRow[] }>(`${EMPLOYERS}/${suffix}`);
}

export function createEmployer(input: EmployerInput): Promise<EmployerRow> {
  return request<EmployerRow>(`${EMPLOYERS}/`, { method: 'POST', body: JSON.stringify(input) });
}

export function updateEmployer(id: string, patch: EmployerInput): Promise<EmployerRow> {
  return request<EmployerRow>(`${EMPLOYERS}/${id}/`, { method: 'PATCH', body: JSON.stringify(patch) });
}

export function fetchEmployerOptions(): Promise<EmployerOptions> {
  return request<EmployerOptions>(`${EMPLOYERS}/options/`);
}

File: backend/learner_api/employers.py
"""JSON API for organisations and employers (no DRF — plain Django views).

Routes (mounted under /learner_api/ in config/urls.py):
    GET   /learner_api/organisations/         -> {count, results:[organisation rows]}
    POST  /learner_api/organisations/         -> create; returns the new row
    GET   /learner_api/organisations/<id>/    -> one organisation
    PATCH /learner_api/organisations/<id>/    -> update; returns the row
    GET   /learner_api/employers/             -> {count, results:[employer rows]}
    POST  /learner_api/employers/             -> create; returns the new row
    GET   /learner_api/employers/<id>/        -> one employer
    PATCH /learner_api/employers/<id>/        -> update; returns the row
    GET   /learner_api/employers/options/     -> dropdown option lists for both forms

An organisation is a company; an employer is a person at one or more of them. The
employer form's "Employer Group" control picks organisations, which is why the
organisation list supports the ?search=/?page= the picker needs.

Authentication follows views.py: write methods require an authenticated staff
or admin session (``@staff_only(writes_only=True)``). Django's CSRF middleware
is exempted because these are JSON endpoints; cross-site protection is the
``X-Requested-With`` header plus the SameSite=Lax session cookie.
"""
from django.db import DatabaseError
from django.views.decorators.csrf import csrf_exempt
from django.http import JsonResponse

from .constants import (
    ORGANISATION_STATUS_CHOICES,
    ORGANISATION_GROUP_TYPE_CHOICES,
    LEVY_PAYER_CHOICES,
    HEALTH_SAFETY_CHOICES,
)
from .mappers import (
    ValidationError,
    to_employer_row,
    to_organisation_row,
    write_employer_fields,
    write_organisation_fields,
)
from login.permissions import staff_only

from .models import Employer, Organisation, StaffUser
from .views import _error, _parse_body, _send_platform_invitation
from login.services import sync_account

# The picker in the reference UI pages ten rows at a time.
PAGE_SIZE = 10


def _resolve_groups(ids):
    """{organisation id: current name} for the ids that actually exist.

    Passed into write_employer_fields, where a missing id is rejected — so this
    doubles as the existence check for the Employer Group selection.
    """
    if not ids:
        return {}
    rows = Organisation.objects.filter(pk__in=ids).values_list("pk", "name")
    return {pk: (name or "").strip() for pk, name in rows}


@csrf_exempt
@staff_only(writes_only=True)
def organisations(request):
    """Organisation profiles — enrolment."Organisations".

    GET supports ?search= (name, case-insensitive), ?status=, and ?page= — the
    employer form's Employer Group picker is a searchable paged table, and the
    list screen reuses the same endpoint.
    """
    if request.method == "GET":
        search = (request.GET.get("search") or "").strip()
        status = (request.GET.get("status") or "").strip()
        try:
            qs = Organisation.objects.all()
            if search:
                qs = qs.filter(name__icontains=search)
            if status:
                qs = qs.filter(status=status)
            qs = qs.order_by("name", "id")
            total = qs.count()

            # ?page= is optional: without it the whole list is returned, which is
            # what a plain "list all organisations" caller wants.
            page = request.GET.get("page")
            if page:
                try:
                    page_no = max(1, int(page))
                except (TypeError, ValueError):
                    return _error("page must be a whole number.", 400)
                start = (page_no - 1) * PAGE_SIZE
                qs = qs[start:start + PAGE_SIZE]
            else:
                page_no = 1

            rows = [to_organisation_row(o) for o in qs]
        except DatabaseError as exc:
            return _error(f"Database error: {exc}", 502)
        return JsonResponse({
            "count": total,
            "page": page_no,
            "pageSize": PAGE_SIZE,
            "results": rows,
        })

    if request.method == "POST":
        try:
            payload = _parse_body(request)
            fields = write_organisation_fields(payload, require_create=True)
        except ValidationError as exc:
            return _error(str(exc), 400)
        try:
            org = Organisation.objects.create(**fields)
        except DatabaseError as exc:
            return _error(f"Database error: {exc}", 502)
        return JsonResponse(to_organisation_row(org), status=201)

    return _error("Method not allowed.", 405)


@csrf_exempt
@staff_only(writes_only=True)
def organisation_detail(request, pk):
    try:
        org = Organisation.objects.get(pk=pk)
    except Organisation.DoesNotExist:
        return _error("Organisation not found.", 404)
    except DatabaseError as exc:
        return _error(f"Database error: {exc}", 502)

    if request.method == "GET":
        return JsonResponse(to_organisation_row(org))

    if request.method in ("PATCH", "PUT"):
        try:
            payload = _parse_body(request)
            fields = write_organisation_fields(payload)
        except ValidationError as exc:
            return _error(str(exc), 400)
        if fields:
            for attr, value in fields.items():
                setattr(org, attr, value)
            try:
                # auto_now on Updated_at only fires when it's in update_fields.
                org.save(update_fields=[*fields.keys(), "updated_at"])
            except DatabaseError as exc:
                return _error(f"Database error: {exc}", 502)
            # A renamed organisation would otherwise leave stale names on the
            # employers pointing at it, since those are denormalised.
            if "name" in fields:
                _resync_group_names(org)
        return JsonResponse(to_organisation_row(org))

    return _error("Method not allowed.", 405)


def _resync_group_names(org):
    """Refresh the denormalised organisation names held on its employers.

    Best-effort: the ids are the real link, so a failure here leaves a stale
    label rather than a broken relationship.
    """
    try:
        name = (org.name or "").strip()
        for emp in Employer.objects.filter(employer_group_ids__contains=org.pk):
            ids = [int(i) for i in (emp.employer_group_ids or [])]
            names = _resolve_groups(ids)
            emp.employer_group_names = [names.get(i, name) for i in ids]
            emp.save(update_fields=["employer_group_names", "updated_at"])
    except DatabaseError:
        pass


@csrf_exempt
@staff_only(writes_only=True)
def employers(request):
    """Employer profiles — enrolment."Employers".

    GET supports ?search= (name or email) and ?organisation= to list the people
    at one organisation.
    """
    if request.method == "GET":
        search = (request.GET.get("search") or "").strip()
        organisation = (request.GET.get("organisation") or "").strip()
        try:
            qs = Employer.objects.all()
            if search:
                from django.db.models import Q
                qs = qs.filter(
                    Q(first_name__icontains=search)
                    | Q(surname__icontains=search)
                    | Q(email__icontains=search)
                )
            if organisation:
                try:
                    # jsonb containment — matches the GIN index on the id array.
                    qs = qs.filter(employer_group_ids__contains=int(organisation))
                except (TypeError, ValueError):
                    return _error("organisation must be a whole number.", 400)
            rows = [to_employer_row(e) for e in qs.order_by("surname", "first_name", "id")]
        except DatabaseError as exc:
            return _error(f"Database error: {exc}", 502)
        return JsonResponse({"count": len(rows), "results": rows})

    if request.method == "POST":
        try:
            payload = _parse_body(request)
            fields = write_employer_fields(
                payload, require_create=True, resolve_groups=_resolve_groups
            )
        except ValidationError as exc:
            return _error(str(exc), 400)
        try:
            emp = Employer.objects.create(**fields)
        except DatabaseError as exc:
            return _error(f"Database error: {exc}", 502)

        row = to_employer_row(emp)
        # Unconditional, as on the learner and staff forms. Nothing is stamped on
        # the row to say so: unlike those two tables, enrolment."Employers" has no
        # "Invite_to_platform" column, and adding one would store a transient
        # action as though it were a property of the person. Whether they were
        # invited is answered by login."Invitations".
        row["invitation"] = _send_platform_invitation(
            request, "employer", emp.id, subject=emp
        )
        return JsonResponse(row, status=201)

    return _error("Method not allowed.", 405)


@csrf_exempt
@staff_only(writes_only=True)
def employer_detail(request, pk):
    try:
        emp = Employer.objects.get(pk=pk)
    except Employer.DoesNotExist:
        return _error("Employer not found.", 404)
    except DatabaseError as exc:
        return _error(f"Database error: {exc}", 502)

    if request.method == "GET":
        return JsonResponse(to_employer_row(emp))

    if request.method in ("PATCH", "PUT"):
        try:
            payload = _parse_body(request)
            fields = write_employer_fields(payload, resolve_groups=_resolve_groups)
        except ValidationError as exc:
            return _error(str(exc), 400)
        if fields:
            for attr, value in fields.items():
                setattr(emp, attr, value)
            try:
                emp.save(update_fields=[*fields.keys(), "updated_at"])
            except DatabaseError as exc:
                return _error(f"Database error: {exc}", 502)
            # The login account holds its own copy of the address an invitation
            # is sent to; correcting it here has to reach that copy.
            if any(field in fields for field in ("email", "full_name")):
                sync_account("employer", emp.pk, subject=emp)
        return JsonResponse(to_employer_row(emp))

    return _error("Method not allowed.", 405)


def employer_options(request):
    """Canonical dropdown lists for the organisation and employer forms.

    Served from the same constants the API validates against, so the two can't
    drift. `owners` is every staff account — an organisation is owned by a member
    of staff, whatever their position, so this is deliberately unfiltered rather
    than restricted to the Caseowner/Admin positions a learner's case owner is.
    """
    if request.method != "GET":
        return _error("Method not allowed.", 405)

    owners = []
    try:
        owners = [
            (u.username or "").strip()
            for u in StaffUser.objects.order_by("username")
            if (u.username or "").strip()
        ]
    except DatabaseError:
        # The forms are still usable with a free-text owner if this lookup fails.
        pass

    return JsonResponse({
        "status": ORGANISATION_STATUS_CHOICES,
        "groupType": ORGANISATION_GROUP_TYPE_CHOICES,
        "levyPayer": LEVY_PAYER_CHOICES,
        "healthAndSafety": HEALTH_SAFETY_CHOICES,
        "owners": owners,
    })
